firebase-firestore/src/main/java/com/google/firebase/firestore/core/Target.java

```diff
@@ -369,6 +373,29 @@ public Direction getKeyOrder() {
     return this.orderBys.get(this.orderBys.size() - 1).getDirection();
   }
 
+  /** Returns the number of segments of a perfect index for this target. */
+  public int getSegmentCount() {
+    Set<FieldPath> fields = new HashSet<>();
+    boolean hasArraySegment = false;
+    for (Filter filter : filters) {
+      for (FieldFilter subFilter : filter.getFlattenedFilters()) {
+        // ARRAY_CONTAINS or ARRAY_CONTAINS_ANY filters must be counted separately. For instance,
+        // it is possible to have an index for "a ARRAY a ASC". Even though these are on the same
+        // field, they should be counted as two separate segments in an index.
+        if (subFilter.getOperator().equals(ARRAY_CONTAINS)
+            || subFilter.getOperator().equals(ARRAY_CONTAINS_ANY)) {
+          hasArraySegment = true;
+        } else {
+          fields.add(subFilter.getField());
+        }
+      }
+    }
+    for (OrderBy orderBy : orderBys) {
+      fields.add(orderBy.getField());
+    }
+    return fields.size() + (hasArraySegment ? 1 : 0);
```

Comment:
We need to figure out or story for __name__. Do you think we should enforce that indices end with __name__ or should we just assume it is always there? I thought we were already ignoring any trailing __name__ elements when an index is configured, which means that we would have to ignore the element here as well. We are, however, not doing this right now so the user would have to end every index with a __name__ component. Let me know if you want to talk about this, but I think it is important to come up with a decision before proceeding much further.

Reply:
yeah, I think all indexes should have an _implicit_ order by `__name__`, and if the user explicitly adds a segment for it, we can drop it and always assume it implicitly exists.

I will update this code to skip the `__name__` fields.